### PR TITLE
Initial Implementation of Padded Tensor

### PR DESCRIPTION
**Pull request by alexanderb14**
This PR introduces the initial implementation of `PaddedTensor`, a Tensor subclass, enabling `reduce-overhead` performance benefits for workloads with dynamic shapes.

## Background and Motivation
Currently, reduce-overhead requires statically shaped models due to constraints in the CUDAGraphs backend. This limitation leads to significant CUDAGraph rerecordings for every shape change, hindering performance. By introducing `PaddedTensor`, we aim to enable reduce-overhead for dynamically shaped models, unlocking performance potential.

## Design
`PaddedTensor` is a Tensor subclass that automatically pads the model to given multipliers, resulting in fewer CUDAGraph recordings.

## Challenges
`PaddedTensor` comes with a design challenge -- the need to propagate the original tensor's shape along with the computation. For `torch.compile`, this is challenging, as it means that the shape computation needs to part of the traced graph as well.

## Design and Implementation
After carefully iterating on several approaches, the approach in this PR represents the original tensor as an inner meta tensor and dispatches ATen and `function` ops on them, along with the data carrying outer padded tensor, which will be static / CUDAGraph'ed. For the inner meta tensor, no data computation is performed.

Contrasting to previous considered approaches, this design combines the benefits:
1. No per-op shape rule maintenance: Unlike representing the original shape as a `torch.Tensor` with manual propagation rules for each ATen op #149140 , the shapes are automatically formed using existing symbolic shape propagation infrastructure.
2. Use of original (symbolic) shape in subsequent operations: This is a requirement for handling masking inputs to operations, i.e. when different operations have different neutral elements. The FakeTensor approach #149241 doesn't align with this requirement, as FakeTensors are created in a separate inner isolated environment that is not part of the traced graph.

## Example: Pointwise operation, with both dimensions padded
```python
def f(a, b):
    return a + b
f = torch.compile(f, fullgraph=True)
multipliers = {0: 16, 1: 16}

for i in range(3, 9):
    a_p = PaddedTensor.from_tensor(torch.randn([i, i]), multipliers)
    b_p = PaddedTensor.from_tensor(torch.randn([i, i]), multipliers)

    y_p = f(a_p, b_p)
    print(y_p)
```

### Lowering 1: No masking
```python
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f32[16, 16]", arg1_1: "f32[s6, s43]", arg2_1: "f32[16, 16]", arg3_1: "f32[s6, s43]"):
         # File: /data/users/alexbrauckmann/pytorch-master/test/inductor/test_padded_tensor.py:133 in f, code: return a + b
        add: "f32[16, 16]" = torch.ops.aten.add.Tensor(arg0_1, arg2_1);  arg0_1 = arg2_1 = None

        # No stacktrace found for following nodes
        empty_strided_default = torch.ops.aten.empty_strided.default((s6, s43), (s43, 1), dtype = torch.float32, device = device(type='meta'))
        return (add, empty_strided_default)
```

### Lowering 2: Masking padded slices
```python
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f32[16, 16]", arg1_1: "f32[s6, s43]", arg2_1: "f32[16, 16]", arg3_1: "f32[s6, s43]"):
         # File: /data/users/alexbrauckmann/pytorch-master/test/inductor/test_padded_tensor.py:133 in f, code: return a + b
        add: "f32[16, 16]" = torch.ops.aten.add.Tensor(arg0_1, arg2_1);  arg0_1 = arg2_1 = None
        sym_size_int: "Sym(s6)" = torch.ops.aten.sym_size.int(arg1_1, 0)
        sym_size_int_1: "Sym(s43)" = torch.ops.aten.sym_size.int(arg1_1, 1);  arg1_1 = None
        full_default: "f32[]" = torch.ops.aten.full.default([], 0.0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
        slice_1: "f32[16 - s6, 16]" = torch.ops.aten.slice.Tensor(add, 0, sym_size_int, 9223372036854775807)
        slice_2: "f32[16 - s6, 16 - s43]" = torch.ops.aten.slice.Tensor(slice_1, 1, sym_size_int_1, 9223372036854775807);  slice_1 = None
        copy: "f32[16 - s6, 16 - s43]" = torch.ops.aten.copy.default(slice_2, full_default);  slice_2 = full_default = None
        slice_3: "f32[16 - s6, 16]" = torch.ops.aten.slice.Tensor(add, 0, sym_size_int, 9223372036854775807)
        slice_scatter: "f32[16 - s6, 16]" = torch.ops.aten.slice_scatter.default(slice_3, copy, 1, sym_size_int_1, 9223372036854775807);  slice_3 = copy = sym_size_int_1 = None
        slice_scatter_1: "f32[16, 16]" = torch.ops.aten.slice_scatter.default(add, slice_scatter, 0, sym_size_int, 9223372036854775807);  add = slice_scatter = sym_size_int = None

        # No stacktrace found for following nodes
        empty_strided_default = torch.ops.aten.empty_strided.default((s6, s43), (s43, 1), dtype = torch.float32, device = device(type='meta'))
        return (slice_scatter_1, empty_strided_default)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Could there be a NestedTensor, but having its storage padded? In this case NestedTensor kernels can be reused as is, and CUDA graph tracing would still be improved because of using only a few underlying padded storage sizes

File: test/inductor/test_padded_tensor.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            self.assertEqual(y_p.shape, (16, 16, 9))
+            self.assertEqual(y_p.original_tensor.shape, (i, i, 9))
+            self.assertEqual(y, y_p.unpad())
+
+
+class ModelTests(TestCase):
+    def setUp(self):
+        super().setUp()
+
+    def test_transformer(self):
```
Comment (etaf) [inline test/inductor/test_padded_tensor.py:168]: Hi， may I suggest we mark this test case as `@requires_cuda` since it use cuda specific API `with torch.device("cuda"):`, `orch._dynamo.config.error_on_recompile`.

File: test/inductor/transformer_model.py [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+"""
+
+import math
+from dataclasses import dataclass
+from typing import Optional
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+
```
Comment (eellison) [inline test/inductor/transformer_model.py:13]: I really don't like these tests that commit the full model... it's just harder to debug when something goes wrong. this should be a test in gpt fast not here
